Snake-Gate Update

dewagter opened the pull request:
- Exclude top and bottom
 - Fix Simulation
 - Only send results on a good gate
 - Simple-Position

dewagter commented:
@guidoAI can you please update this

dewagter commented on sw/airborne/modules/computer_vision/detect_gate.c:
```diff
@@ -71,8 +74,9 @@ PRINT_CONFIG_VAR(DETECT_GATE_Y_MIN)
+// now 42 in sim, 66 before
 #ifndef DETECT_GATE_U_MIN
-#define DETECT_GATE_U_MIN 66
+#define DETECT_GATE_U_MIN 42
```
This is only to catch cases where the colors are not defined. If we add a warning, that the user should define them in the airframe?

kirkscheper replied:
No, just remove them, I think the user should have to define them, there should be no default. This is mission specific not module specific.

kirkscheper replied:
We all know orange is never really orange, especially in YUV.

dewagter commented on sw/airborne/modules/computer_vision/detect_gate.c:
```diff
@@ -168,24 +208,67 @@ struct image_t *detect_gate_func(struct image_t *img)
-      drone_position = get_world_position_from_image_points(best_gate.x_corners, best_gate.y_corners, world_corners,
-                       n_corners,
-                       DETECT_GATE_CAMERA.camera_intrinsics, cam_body);
+      static bool simple_position = false;
```
Yes, this provides 2 reprojection model options: Full PnP or simple FOV

dewagter commented:
- removed default color as this must be set by all applications
 - explained both reprojection models in XML and added define
 - swapped the (commented) SIMULATE define by a hopefully more descriptive CAMERA_ROTATED_90DEG

kirkscheper commented on sw/airborne/modules/computer_vision/detect_gate.c:
```diff
@@ -128,9 +119,19 @@ volatile float detect_gate_z;
 static pthread_mutex_t gate_detect_mutex;            ///< Mutex lock fo thread safety
 
+// Module variables
+struct vision_relative_position_struct {
```
Seems a little redundant to name the struct struct, I think the convention is to use _t

guidoAI commented on sw/airborne/modules/computer_vision/detect_gate.c:
```diff
@@ -168,45 +194,116 @@ struct image_t *detect_gate_func(struct image_t *img)
+        drone_position = get_world_position_from_image_points(best_gate.x_corners, best_gate.y_corners, world_corners,
+                               n_corners, DETECT_GATE_CAMERA.camera_intrinsics, cam_body);
+      }
+      printf("Position drone: (%f, %f, %f)\n", drone_position.x, drone_position.y, drone_position.z);
```
yes, you're right

guidoAI commented on sw/airborne/modules/computer_vision/detect_gate.c:
```diff
@@ -168,45 +194,116 @@ struct image_t *detect_gate_func(struct image_t *img)
-  static int32_t cnt = 0;
   pthread_mutex_lock(&gate_detect_mutex);
   if (detect_gate_has_new_data) {
+    //printf("Sending data!\n");
```
check

guidoAI commented on sw/airborne/modules/computer_vision/snake_gate_detection.c:
```diff
@@ -331,7 +338,25 @@ int snake_gate_detection(struct image_t *img, int n_samples, int min_px_size, fl
+        ratio = limit_ratio + 0.1;
+      }
+
+      //printf("Gate with surface area: %f, quality %f, n_sides %d, and ratio %f\n", sz1g*sz2g, gates_c[gate_nr].quality, gates_c[gate_nr].n_sides, ratio);
```
removed

guidoAI commented on sw/airborne/modules/computer_vision/snake_gate_detection.c:
```diff
@@ -331,7 +338,25 @@ int snake_gate_detection(struct image_t *img, int n_samples, int min_px_size, fl
+
+      // if (gates_c[gate_nr].n_sides >= min_n_sides && gates_c[gate_nr].quality > best_gate->quality) {
+      if (sz1g*sz2g > sz1*sz2 && gates_c[gate_nr].quality > min_gate_quality * 2 && gates_c[gate_nr].n_sides >= min_n_sides && ratio <= limit_ratio) {
+        //printf("Becomes the best gate!\n");
```
Now added a comment to explain the difference between the if statements.
